File: day4/main.py



if __name__ == '__main__':
    lines = [line.strip() for line in open("test.txt")]

    key = "XMAS"
    key_backward = key[::-1]
    total = 0
    # horizontal count
    for line in lines:
        # how many times key appears in line
        total += line.count(key)
        total += line.count(key_backward)
        
    print('after horizontal search', total)
    # vertical search
    # stitch strings vertically
    chars = [list(line) for line in lines]
    vertical = list(zip(*chars))
    vertical = ["".join(line) for line in vertical]

    for line in vertical:
        total += line.count(key)
        total += line.count(key_backward)

    print("after vertical search", total)

    for r in range(len(lines[:-3])):
        for c in range(len(lines[0])-3):
            # right down
            if lines[r][c] + lines[r+1][c+1] + lines[r+2][c+2] + lines[r+3][c+3] == key:
                total += 1
            if lines[r][c] + lines[r+1][c+1] + lines[r+2][c+2] + lines[r+3][c+3] == key_backward:
                total += 1
            # left down
            if lines[r][c+3] + lines[r+1][c+2] + lines[r+2][c+1] + lines[r+3][c] == key:
                total += 1
            if lines[r][c+3] + lines[r+1][c+2] + lines[r+2][c+1] + lines[r+3][c] == key_backward:
                total += 1

    print("final", total)
    print()

    # part 2

    key = "MAS"
    key_backward = key[::-1]
    total = 0
    for r in range(len(lines[:-2])):
        for c in range(len(lines[0]) - 2):
            right_down = lines[r][c] + lines[r+1][c+1] + lines[r+2][c+2]
            if right_down == key or right_down == key_backward:
                left_down = lines[r+2][c] + lines[r+1][c+1] + lines[r][c+2]
                if left_down == key or left_down == key_backward:
                    total += 1


    print('part 2', total)

File: day4/main.cpp
#include <cmath>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>
#include <iostream>

int part1(const std::vector<std::string> &lines) {
    int total = 0;
    std::string key = "XMAS";
    std::string key_backwards = "SAMX";

    // horizontal search
    for (int r = 0; r < lines.size(); r++) {
        std::string line = lines[r];
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
    }

    // vertical search
    for (int c = 0; c < lines[0].size(); c++) {
        std::string line = "";
        for (int r = 0; r < lines.size(); r++) {
            line += lines[r][c];
        }
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            std::cout << "found match: " << line << std::endl;
            total++;
            pos += 1;
        }
        pos = 0;
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            std::cout << "found match: " << line << std::endl;
            total++;
            pos += 1;
        }
    }

    // bottom left corner
    for (int r = 0; r < lines.size(); r++) {
        std::string line = "";
        for (int i = 0; r + i < lines.size(); i++) {
            line += lines[r + i][i];
        }
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        pos = 0;
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
    }
    // top right corner
    for (int c = 1; c < lines[0].size(); c++) {
        std::string line = "";
        for (int i = 0; c + i < lines[0].size(); i++) {
            line += lines[i][c + i];
        }
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        pos = 0;
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
    }

    // bottom right corner
    for (int r = 0; r < lines.size(); r++) {
        std::string line = "";
        for (int i = 0; r + i < lines.size(); i++) {
            line += lines[r + i][lines[0].size() - 1 - i];
        }
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        pos = 0;
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        
    }
    // top left corner
    for (int c = 0; c < lines[0].size() - 1; c++) {
        std::string line = "";
        for (int i = 0; c - i >= 0; i++) {
            line += lines[i][c - i];
        }
        int pos = 0;
        while ((pos = line.find(key, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        pos = 0;
        while ((pos = line.find(key_backwards, pos)) != std::string::npos) {
            total++;
            pos += 1;
        }
        std::cout << line << std::endl;
    }

    return total;
}

int part2(const std::vector<std::string> &lines) {
    return 0;
}

int main() {
    std::ifstream file("test.txt");
    if (!file.is_open()) {
        std::cerr << "Error: could not open file" << std::endl;
        return 1;
    }

    std::vector<std::string> lines(
        std::istream_iterator<std::string>{file},
        std::istream_iterator<std::string>()
    );

    std::cout << "Part1: " << part1(lines) << std::endl;
    std::cout << "Part2: " << part2(lines) << std::endl;
}